### PR TITLE
[flang][OpenMP] Implicitly map allocatable record fields

### PR DESCRIPTION
**Pull request by ergawy**
This is a starting PR to implicitly map allocatable record fields.

This PR contains the following changes:
1. Re-purposes some of the utils used in `Lower/OpenMP.cpp` so that
   these utils work on the `mlir::Value` level rather than the
   `semantics::Symbol` level. This takes one step towards to enabling
   MLIR passes to more easily do some lowering themselves (e.g. creating
   `omp.map.bounds` ops for implicitely caputured data like this PR
   does).
2. Adds support for implicitely capturing and mapping allocatable fields
   in record types.

There is quite some distant to still cover to have full support for
this. I added a number of todos to guide further development.

Co-authored-by: Andrew Gozillon <andrew.gozillon@amd.com>

### CODE REVIEW

File: flang/lib/Lower/DirectivesCommon.h
Comment (ergawy) [inline flang/lib/Lower/DirectivesCommon.h:632]: This is a dead condition since we call `symAddr.getDefiningOp()` above. I think we should assert instread.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (agozillon) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:523]: Might need to be a tad wary of how this works with use_device_addr/ptr, as they're both map info holders, and soon with Pranav's and your work I believe private may also be, and I don't think they have the same implicit connotations regular map does. We may also eventually need to be careful of this with declare mappers as well, as they're explicit user defined mappings that shouldn't have the implicit behavior but that's a problem for another day/PR when that work is further along :-)

File: offload/test/offloading/fortran/implicit-record-field-mapping.f90
Comment (agozillon) [inline offload/test/offloading/fortran/implicit-record-field-mapping.f90:17]: Excellent start! :-) 

Will likely need to consider the cases where we have an allocatable derived type, with allocatable members (or regular members), and cases where we have multiple nested allocatable derived types within each other. Basically just mixes of various allocatable/pointers etc. Doesn't need to work in the first PR iteration, but at the very least we need to make sure we don't blow up the compiler or runtime mappings by accident with the new implicit mappings.

Although, I think if you can run check-offload, this should all be covered as-is, but please do make sure check-offload runs correctly for you (e.g. intentionally fail a test locally and run to make sure it does in fact test them all), there's annoying cases where check-offload doesn't actually run the fortran offload tests even when Flang and all it's relevant components have been compiled! Unsure why unfortunately, likely related to environment setup.

But still might want to prevent the generation of these in nested cases if you're going to leave them as a TODO for now, just to prevent any possible erroneous cases that aren't caught by check-offload and to prevent unnecessary mappings that may not be correct!

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (agozillon) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:580]: for the whole recursive generation / gather of indices, you may be able to borrow / tweak (improve on ;-)) the explicit map code that does something similar in the function: "`createParentSymAndGenIntermediateMaps`"

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (agozillon) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:614]: It's worth noting that even implicit maps that aren't directly utilised in the TargetOp's region get inserted into the BlockArgument list of TargetOp, so please do make sure that's still occurring to maintain consistency (especially as I believe the BlockArgInterface is somewhat sensitive to it, could be incorrect however). For TargetDataOp that's not currently the case, which makes things a little weird as they'd ideally work the same way :-) Perhaps one day soon. 

Would love to see a test showcasing the pre/post MapInfoFinalization pass FIR/HLFIR in the final PR when it's ready, similar to the other tests for this pass! I'll leave it up to you if you'd prefer to include it in the same test file as the rest or another, as the current test file is verifying slightly different behavior :-)

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (agozillon) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:605]: should probably take the capture kind from the original operation as well, to keep it inline with the map type attribute, may save some possible weirdness :-) 

But taking the map type from the parent does raise the interesting question of if the implicit mapping should take on the specified map type behavior of it's parent or if it should default to the regular implicit map behaviour of it's type. Not too sure on that one, the spec or one of our local specification gurus may have some insight into that.

**Review by agozillon — COMMENTED** [review]
This LGTM so far, thank you very much for the work, it looks excellent so far. Left some comments here and there mainly just little notes to hopefully help out / forewarn of certain things :-) 

I think the approach of doing it as an opt/transformation pass is fine personally, and I believe at least Slava from the OpenACC team originally mentioned an approach like this would be ideal for the implicit mapping. The only issue I can see someone bringing up is that we now have a split of implicit mapping between the opt passes and the initial lowering, I personally don't mind though and think it's fine, we can always move the front end lowering down to here at some point if it's a cleaner approach for the implicit mapping and it doesn't hinder anything else. And I think the opt pass approach is beneficial as we don't need to recreate the wheel for the do concurrent work, or any other work that may depend on this behavior.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:584]: Thanks for the pointer. Also added a `todo` for a later PR.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:523]: Added a `todo`.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:505]: Add a `todo`.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp]: Added an MLIR toMLIR test.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (ergawy) [reply flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:609]: Maybe @mjklemm can provide some guidance here?

Comment (mjklemm) [reply flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:609]: Yes, I think that's the right behavior.  However, we need to bear in mind that a user might have used `declare mapper` to define a mapper for a derived type.  In that case, we might need to honor the user-defined mapper instead.

**Comment by ergawy** [comment]
Ping 🔔! Would be nice to take a look when you folks have a chance 🙏! There is still quite some work to here in follow up PRs.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (luporl) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:499]: I guess this won't work for all cases, as when `underlyingType` is an array of records, for instance.
It could be a TODO.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (luporl) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:531]:
```suggestion
          // the corresponding parameter to the field in the called function.
```

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
Comment (luporl) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:537]:
```suggestion
        llvm::SmallVector<int64_t> fieldIndices;
```

File: flang/test/Transforms/omp-map-info-finalization-implicit-field.fir
Comment (luporl) [inline flang/test/Transforms/omp-map-info-finalization-implicit-field.fir:6]:
```suggestion
!record_t = !fir.type<_QFTrecord_t{
```

File: offload/test/offloading/fortran/implicit-record-field-mapping.f90
Comment (luporl) [inline offload/test/offloading/fortran/implicit-record-field-mapping.f90:14]:
```suggestion
  type record_t
```

**Review by agozillon — APPROVED** [review]
LGTM after the last change :-) thank you very much for the patch!

**Comment by ergawy** [comment]
Ping 🔔! Please let me know if there are objections to merging this PR. Since a lot of people are on vacation, I am waiting a bit more before merging to give more time.

**Review by mjklemm — APPROVED** [review]
LGTM

**Comment by kazutakahirata** [comment]
@ergawy I'm seeing the same error as the buildbots above.  Would you mind taking a look?  Thanks!

**Comment by ergawy** [comment]
Created a revert PR for now: https://github.com/llvm/llvm-project/pull/120360 while investigating the issue.

**Pull request merged**
